lib/Dialect/FIRRTL/InnerSymbolTable.cpp

```diff
@@ -0,0 +1,152 @@
+//===- InnerSymbolTable.cpp - InnerSymbolTable and InnerRef verification --===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements InnerSymbolTable and verification for InnerRef's.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/FIRRTL/FIRRTLOpInterfaces.h"
+#include "mlir/IR/Threading.h"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace circt {
+namespace firrtl {
+
+//===----------------------------------------------------------------------===//
+// InnerSymbolTable
+//===----------------------------------------------------------------------===//
+
+InnerSymbolTable::InnerSymbolTable(Operation *op) {
+  assert(op->hasTrait<OpTrait::InnerSymbolTable>() &&
+         "expected operation to have InnerSymbolTable trait");
+  assert(op->getNumRegions() == 1 &&
+         "expected operation to have a single region");
+  assert(llvm::hasSingleElement(op->getRegion(0)) &&
+         "expected operation to have a single block");
```

Comment:
I think we should enforce the single region and single blocks as a a part of the OpTrait verifier so we don't have to check them here

Reply:
Works for me.  Same for the `verifyInnerRefs` checks, since while that code will crash if those conditions aren't met (Operation with no regions), it really should only ever be called by the trait verifier.